tests/integration/security/rbac/util/rbac_util.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"testing"
+
+	"istio.io/istio/pkg/test/framework/components/apps"
+	"istio.io/istio/pkg/test/util/connection"
+	"istio.io/istio/tests/util"
+)
+
+type TestCase struct {
+	Request       connection.Connection
+	ExpectAllowed bool
+	Jwt           string
+	RejectionCode string
+}
+
+// CheckRBACRequest checks if a request is successful under RBAC policies.
+// Under RBAC policies, a request is consider successful if:
+// * If the policy is deny:
+// *** For HTTP: response code is same as the tc.RejectionCode.
+// *** For TCP: EOF error
+// * If the policy is allow:
+// *** Response code is 200
+func CheckRBACRequest(tc TestCase) error {
+	req := tc.Request
+	ep := req.To.EndpointForPort(req.Port)
+	if ep == nil {
+		return fmt.Errorf("cannot get upstream endpoint for connection test %v", req)
+	}
+
+	headers := make(http.Header)
+	if len(tc.Jwt) > 0 {
+		headers.Add("Authorization", "Bearer "+tc.Jwt)
+	}
+
+	resp, err := req.From.Call(ep, apps.AppCallOptions{Protocol: req.Protocol, Path: req.Path, Headers: headers})
+	if err != nil && !strings.Contains(err.Error(), "EOF") {
+		return fmt.Errorf("connection error with %v", err)
+	}
+
+	if tc.ExpectAllowed {
+		if !(len(resp) > 0 && resp[0].Code == connection.AllowHTTPRespCode) {
+			return fmt.Errorf("%s to %s:%d%s using %s: expected allow, actually deny",
+				req.From.Name(), req.To.Name(), req.Port, req.Path, req.Protocol)
+		}
+	} else {
+		if req.Port == connection.TCPPort {
+			if !strings.Contains(err.Error(), "EOF") {
+				return fmt.Errorf("%s to %s:%d%s using %s: expected deny with EOF error, actually %v",
+					req.From.Name(), req.To.Name(), req.Port, req.Path, req.Protocol, err)
+			}
+		} else {
+			if !(len(resp) > 0 && resp[0].Code == tc.RejectionCode) {
```

Comment:
Hard-coding the rejection code in the function makes the function CheckRBACRequest() less extensible and hard to understand (an user for this function has to read the entire funciton to understand the hardcoded rejection code). Taking the rejection code as an input parameter is more extensible and makes the function easier to use.

Reply:
It's already called `CheckRBACRequest()` and all the caller are passing the same `RejectionCode`, The expected deny code for RBAC is always 403, so in what case we will need to extend this behavior?

Reply:
Example cases of extending can be 1) the rejection code changes, 2)extending the function for checking another response code. 
Hard-coding may satisfy current requirements but makes the code hard to read and hard to extend. It is better software engineering practice to write code that is readable and extensible.

Reply:
You pass all the same value of `RejectionCode` in each case, some of test case use it and some don't, this is confusing as you don't know how will it affect the test.
Second, if we really need to change the "403" response code (I doubt this), we can simply change it in `CheckRBACRequest()`, as compared to today, you have to change it everywhere in each test cases.
Last, if this is really needed (again I doubt this), it should be merged with the `ExpectAllowed`, the point is we really need 1 field to indicate the expected outcome of the test, which makes it much easier to understand the test by only looking at each test case, instead of digging into the `CheckRBACRequest()`

Reply:
CheckRBACRequest() makes a request and results in a response code. The response may be allow or rejection, but not necessarily allow, thus a rejection code is specified by the test cases. 
If a hardcoded rejection code was used, it is also in CheckRBACRequest()  for every case. 

The test cases use a macro to specify their rejection code, so changing the rejection code of tests cases is a one-line change. In contrast, if hard-coding the rejection code in CheckRBACRequest(), a developer needs to read into the function, understand its implementation, change the hardcoded rejection code, which may break the existing tests that depend on the original hardcoded rejection code. Moreover, the hardcoded rejection code does not support the case that some tests have rejection code 1 while some other tests have rejection code 2.

Regarding to merge this field (i.e., the rejection code) with the ExpectAllowed field, one is rejection code for the http request where the other field is expected test case outcome; they are two disjoint concepts and merging them is confusing.